sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -121,6 +121,19 @@ object ScalaReflection extends ScalaReflection {
     case _ => false
   }
 
+  def classForNativeTypeOf(dt: DataType): Class[_] = dt match {
```

Comment:
Shouldn't this match `CodeGenerator.boxedType` in terms of functionality? See my previous comments, but I am also missing complex types (struct, map & array).

Reply:
IIUC this function would be better to only handles the `native` types that have the same format between the Spark SQL internals formats and externals (struct, map, and array have different formats between them). https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala#L209 So, this function needs to handle more wider types than `CodeGenerator.boxedType`.

Reply:
There is a difference between how we implement an expression and how we use an expression. In this case the implementations should behave the same, and not only in the context in which it is being used.

Reply:
ok, I'll brush up the code.

Reply:
@hvanhovell How about the latest commit?